### PR TITLE
Extend compatibility of MemoryWatcher to Windows

### PR DESCRIPTION
**Pull request by TechNick6425**
MemoryWatcher only worked on *nix before, but I think I made it work on Windows. Uses a Named Pipe to communicate over "\\.\dolphin-emu\mem-watch"

### CODE REVIEW

File: Source/Core/Core/MemoryWatcher.cpp [hunk 1]
```diff
@@ -62,12 +77,26 @@ void MemoryWatcher::ParseLine(const std::string& line)
...
 {
+#ifdef __unix__
 	memset(&m_addr, 0, sizeof(m_addr));
 	m_addr.sun_family = AF_UNIX;
 	strncpy(m_addr.sun_path, path.c_str(), sizeof(m_addr.sun_path) - 1);
 
 	m_fd = socket(AF_UNIX, SOCK_DGRAM, 0);
 	return m_fd >= 0;
+#elif defined(_WIN32)
+	pipe = CreateNamedPipe(L"\\\\.\\dolphin-emu\\mem-watch",
```
Comment (Alcaro) [inline Source/Core/Core/MemoryWatcher.cpp]: C++11 raw strings would cut the amount of backslashes in half.

File: Source/Core/Core/MemoryWatcher.cpp [hunk 2]
```diff
@@ -62,12 +77,26 @@ void MemoryWatcher::ParseLine(const std::string& line)
...
+	pipe = CreateNamedPipe(L"\\\\.\\dolphin-emu\\mem-watch", 
+		PIPE_ACCESS_OUTBOUND,
+		PIPE_TYPE_BYTE | PIPE_WAIT,
+		1,
+		0,
+		0,
+		0,
+		NULL);
+
+	if(pipe == NULL | pipe == INVALID_HANDLE_VALUE) return false;
```
Comment (Alcaro) [inline Source/Core/Core/MemoryWatcher.cpp]: Use || for booleans. | is a different operator.

And [docs](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365150%28v=vs.85%29.aspx) say NULL is not the error return, so no point checking that at all.

File: Source/Core/Core/MemoryWatcher.cpp [hunk 3]
```diff
@@ -2,12 +2,22 @@
...
+#ifdef __unix__
+
 #include <fstream>
 #include <iostream>
 #include <sstream>
 #include <unistd.h>
 
+#elif defined(_WIN32)
+
+#include <iostream>
```
Comment (Alcaro) [inline Source/Core/Core/MemoryWatcher.cpp]: If it's needed for all platforms, why is it in the ifdef at all?

I'd recommend moving out the other C++ headers (fstream, sstream) too. IIRC, MSVC includes a huge pile of other headers from iostream, I wouldn't recommend relying on that.

File: Source/Core/Core/MemoryWatcher.h [hunk 4]
```diff
@@ -28,6 +37,7 @@ class MemoryWatcher final
 private:
 	bool LoadAddresses(const std::string& path);
 	bool OpenSocket(const std::string& path);
+	void CloseSocket();
```
Comment (Alcaro) [inline Source/Core/Core/MemoryWatcher.h]: This one isn't defined anywhere. Or used.

File: Source/Core/Core/MemoryWatcher.cpp [hunk 5]
```diff
@@ -62,12 +77,26 @@ void MemoryWatcher::ParseLine(const std::string& line)
...
 	return m_fd >= 0;
+#elif defined(_WIN32)
+	pipe = CreateNamedPipe(L"\\\\.\\dolphin-emu\\mem-watch", 
+		PIPE_ACCESS_OUTBOUND,
+		PIPE_TYPE_BYTE | PIPE_WAIT,
+		1,
+		0,
+		0,
+		0,
+		NULL);
```
Comment (Alcaro) [inline Source/Core/Core/MemoryWatcher.cpp]: Style guide says nullptr.

File: Source/Core/Core/MemoryWatcher.h [hunk 6]
```diff
@@ -38,9 +48,17 @@ class MemoryWatcher final
...
 	std::atomic_bool m_running{false};
 
+#ifdef __unix__
+
 	int m_fd;
 	sockaddr_un m_addr;
 
+#elif defined(_WIN32)
+
+	HANDLE pipe;
```
Comment (Alcaro) [inline Source/Core/Core/MemoryWatcher.h]: Style guide says m_ on member variables.

**Comment by TechNick6425** [comment]
I'll fix those things you mentioned.

File: Source/Core/Core/MemoryWatcher.cpp (hunk 2 above)
Comment (JosJuice) [inline Source/Core/Core/MemoryWatcher.cpp]: There must be a space after `if`, both here and in other places.

Comment (JosJuice) [inline Source/Core/Core/MemoryWatcher.cpp]: `return false;` must be on a new line. (Though in this case you can simply have `return pipe != INVALID_HANDLE_VALUE;`)

File: Source/Core/Core/MemoryWatcher.cpp [hunk 7]
```diff
@@ -107,6 +146,12 @@ void MemoryWatcher::WatcherThread()
 					0,
 					reinterpret_cast<sockaddr*>(&m_addr),
 					sizeof(m_addr));
+#elif defined(_WIN32)
+				if(!WriteFile(pipe, message, sizeof(char) * strlen(message), NULL, NULL))
```
Comment (lioncash) [inline Source/Core/Core/MemoryWatcher.cpp]: This line alone tells me you haven't tested this code. Passing a std::string to strlen isn't going to work.

File: Source/Core/Core/MemoryWatcher.cpp [hunk 8]
```diff
@@ -107,6 +146,12 @@ void MemoryWatcher::WatcherThread()
 					0,
 					reinterpret_cast<sockaddr*>(&m_addr),
 					sizeof(m_addr));
+#elif defined(_WIN32)
+				if(!WriteFile(m_pipe, message, sizeof(char) * strlen(message), NULL, NULL))
```
Comment (lioncash) [inline Source/Core/Core/MemoryWatcher.cpp]: This line alone tells me you haven't even tested this code. strlen called on a std::string isn't going to work.

**Comment by lioncash** [comment]
Please test your code before making a PR for it in the future. 

Also, this should be implemented as an interface. The MemoryWatcher class should become an interface class, and then be inherited from for each supporting platform. This keeps out unnecessary ifdef soup.

**Pull request closed without merging**